File: server/src/main/java/com/krillsson/sysapi/core/history/MetricsHistoryManager.java
package com.krillsson.sysapi.core.history;

import com.google.common.eventbus.EventBus;
import com.krillsson.sysapi.config.HistoryConfiguration;
import com.krillsson.sysapi.core.domain.cpu.CpuLoad;
import com.krillsson.sysapi.core.domain.drives.DriveLoad;
import com.krillsson.sysapi.core.domain.gpu.GpuLoad;
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceLoad;
import com.krillsson.sysapi.core.domain.system.SystemLoad;
import oshi.hardware.GlobalMemory;

import java.util.List;
import java.util.stream.Collectors;

public class MetricsHistoryManager extends HistoryManager {


    public MetricsHistoryManager(HistoryConfiguration configuration, EventBus eventBus) {
        super(configuration, eventBus);
    }

    public List<History.HistoryEntry<CpuLoad>> cpuLoadHistory() {
        return systemLoadHistory().stream()
                .map(e -> new History.HistoryEntry<CpuLoad>(e.date, e.value.getCpuLoad()))
                .collect(Collectors.toList());
    }

    public List<History.HistoryEntry<SystemLoad>> systemLoadHistory() {
        return getHistory();
    }

    public List<History.HistoryEntry<List<DriveLoad>>> driveLoadHistory() {
        return systemLoadHistory().stream()
                .map(e -> new History.HistoryEntry<List<DriveLoad>>(e.date, e.value.getDriveLoads()))
                .collect(Collectors.toList());
    }

    public List<History.HistoryEntry<List<GpuLoad>>> gpuLoadHistory() {
        return systemLoadHistory().stream()
                .map(e -> new History.HistoryEntry<List<GpuLoad>>(e.date, e.value.getGpuLoads()))
                .collect(Collectors.toList());
    }

    public List<History.HistoryEntry<GlobalMemory>> memoryHistory() {
        return systemLoadHistory().stream()
                .map(e -> new History.HistoryEntry<GlobalMemory>(e.date, e.value.getMemory()))
                .collect(Collectors.toList());
    }

    public List<History.HistoryEntry<List<NetworkInterfaceLoad>>> networkInterfaceLoadHistory() {
        return systemLoadHistory().stream()
                .map(e -> new History.HistoryEntry<List<NetworkInterfaceLoad>>(
                        e.date,
                        e.value.getNetworkInterfaceLoads()
                ))
                .collect(Collectors.toList());
    }
}

File: server/src/main/java/com/krillsson/sysapi/core/history/MetricsHistoryManager.kt
package com.krillsson.sysapi.core.history

import com.google.common.eventbus.EventBus
import com.krillsson.sysapi.config.HistoryConfiguration
import com.krillsson.sysapi.core.domain.cpu.CpuLoad
import com.krillsson.sysapi.core.domain.drives.DriveLoad
import com.krillsson.sysapi.core.domain.gpu.GpuLoad
import com.krillsson.sysapi.core.domain.history.HistoryEntry
import com.krillsson.sysapi.core.domain.history.SystemHistoryEntry
import com.krillsson.sysapi.core.domain.memory.MemoryLoad
import com.krillsson.sysapi.core.domain.network.NetworkInterfaceLoad
import java.time.OffsetDateTime
import java.util.stream.Collectors

class MetricsHistoryManager(configuration: HistoryConfiguration, eventBus: EventBus, historyRepository: HistoryRepository) :
    HistoryManager(configuration, eventBus, historyRepository) {
    fun cpuLoadHistory(
        fromDate: OffsetDateTime?,
        toDate: OffsetDateTime?
    ): List<HistoryEntry<CpuLoad>> {
        return systemLoadHistory(fromDate, toDate).stream()
            .map { e: SystemHistoryEntry ->
                HistoryEntry(
                    e.date,
                    e.value.cpuLoad
                )
            }
            .collect(Collectors.toList())
    }

    fun systemLoadHistory(
        fromDate: OffsetDateTime?,
        toDate: OffsetDateTime?
    ): List<SystemHistoryEntry> {
        return getHistoryLimitedToDates(fromDate, toDate)
    }

    fun driveLoadHistory(
        fromDate: OffsetDateTime?,
        toDate: OffsetDateTime?
    ): List<HistoryEntry<List<DriveLoad>>> {
        return systemLoadHistory(fromDate, toDate).stream()
            .map { e: SystemHistoryEntry ->
                HistoryEntry(
                    e.date,
                    e.value.driveLoads
                )
            }
            .collect(Collectors.toList())
    }

    fun gpuLoadHistory(
        fromDate: OffsetDateTime?,
        toDate: OffsetDateTime?
    ): List<HistoryEntry<List<GpuLoad>>> {
        return systemLoadHistory(fromDate, toDate).stream()
            .map { e: SystemHistoryEntry ->
                HistoryEntry(
                    e.date,
                    e.value.gpuLoads
                )
            }
            .collect(Collectors.toList())
    }

    fun memoryHistory(
        fromDate: OffsetDateTime?,
        toDate: OffsetDateTime?
    ): List<HistoryEntry<MemoryLoad>> {
        return systemLoadHistory(fromDate, toDate).stream()
            .map { e: SystemHistoryEntry ->
                HistoryEntry(
                    e.date,
                    e.value.memory
                )
            }
            .collect(Collectors.toList())
    }

    fun networkInterfaceLoadHistory(
        fromDate: OffsetDateTime?,
        toDate: OffsetDateTime?
    ): List<HistoryEntry<List<NetworkInterfaceLoad>>> {
        return systemLoadHistory(fromDate, toDate).stream()
            .map { e: SystemHistoryEntry ->
                HistoryEntry(
                    e.date,
                    e.value.networkInterfaceLoads
                )
            }
            .collect(Collectors.toList())
    }
}